### PR TITLE
Make IPSET integration optional and non-fatal

### PR DESCRIPTION
**Pull request by jumpsmm7**
### Motivation
- Users reported AdGuardHome failing to start when the managed IPSET integration produced no data or when IPSET was not used on their router, causing monitor restart loops.
- The installer needed a way to enable/disable the managed IPSET integration from the UI so installations without IPSET can opt out and avoid startup failures.

### Description
- Add an installer/config setting `ADGUARD_IPSET` with interactive controls (`check_ipset`) and a new menu entry `8) Enable/Disable AdGuardHome IPSET Integration` to enable or disable managed IPSET integration via the installer menu and reconfigure flow.
- Make IPSET integration opt-in/opt-out at runtime with `IPSet_Enabled()` and gate `IPSet_Setup`, `IPSet_Setup_For_Start`, and `IPSet_Refresh` on that check so a `NO` value disables managed IPSET behavior while preserving `ipset.user`.
- Make IPSET startup preparation non-fatal by continuing AdGuardHome startup when `IPSet_Setup_For_Start` fails and logging a warning instead of aborting startup and triggering monitor restarts.
- Change `IPSet_Refresh_Locked` to treat an empty generated mapping set as no-data: it removes the temporary/empty files, calls `IPSet_Disable_Managed` (clearing the managed `dns.ipset_file`), and does not prevent AdGuardHome from starting; later refreshes will re-enable managed integration if mappings appear.
- Update documentation (`README.md`) to describe the toggle, backward-compatible default, empty-mapping behavior, and that `ipset.user` is preserved.
- Add `tests/ipset-empty-rules.sh`, update `tests/start-adguardhome-lifecycle.sh` and `tests/ipset-version-gate.sh` to cover disabled integration, empty mapping behavior, and startup fallback, and register the new test in `tools/code-quality.sh`.

### Testing
- Ran `sh tools/check-md5.sh` and it passed successfully.
- Ran all regression scripts with `for test in tests/*.sh; do sh "$test"; done` and all tests passed, including the new `tests/ipset-empty-rules.sh` and updated lifecycle/version tests.
- Ran a full script syntax check with `sh -n` over the repository scripts and `sh tools/code-quality.sh`; all project checks passed, however `shellcheck` and `shfmt` are not available in this environment (reported as warnings), so those static/format checks were not executed here.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a2c1ee12680832982ae022b6495476a)

### CODE REVIEW

File: AdGuardHome.sh [hunk 1]
```diff
@@ -1783,6 +1784,15 @@ IPSet_Refresh_Locked() {
 		return 1
 	fi
 	rm -f "${RAW_TEMP_FILE}"
+	if ! awk '!/^[[:space:]]*(#|$)/ { found = 1; exit } END { exit !found }' "${TEMP_FILE}"; then
+		rm -f "${RAW_TEMP_FILE}" "${TEMP_FILE}" "${IPSET_FILE}"
+		if ! IPSet_Disable_Managed; then
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @codex resolve that P1

File: installer [hunk 2]
```diff
@@ -1836,7 +1850,7 @@ menu() {
 			PTXT "${INFO} Choose what you want to do:" \
 				"  1) Install/Update AdGuardHome" \
 				"  2) Uninstall"
-			if { [ -d "${TARG_DIR}" ] && [ -f "${AGH_FILE}" ]; }; then { PTXT "  3) Change Username/Password" "  4) Reconfigure" "  5) Enable/Disable AMTM updates" "  6) Enable/Disable AdGuardHome Local Cache" "  7) Switch AdGuardHome Branch" "  b) Backup"; }; fi
+			if { [ -d "${TARG_DIR}" ] && [ -f "${AGH_FILE}" ]; }; then { PTXT "  3) Change Username/Password" "  4) Reconfigure" "  5) Enable/Disable AMTM updates" "  6) Enable/Disable AdGuardHome Local Cache" "  7) Switch AdGuardHome Branch" "  8) Enable/Disable AdGuardHome IPSET Integration" "  b) Backup"; }; fi
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer:1872]: @codex resolve that P1

File: AdGuardHome.sh [hunk 3]
```diff
@@ -1783,6 +1784,15 @@ IPSet_Refresh_Locked() {
...
+			return 1
+		fi
+		IPSET_REFRESH_CHANGED="1"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @codex resolve that P2

File: installer [hunk 4]
```diff
@@ -1897,6 +1911,15 @@ menu() {
 				"${INFO} Before you choose to proceed, it is recommended to perform a backup."
 			if read_yesno "Do you want to proceed?"; then { choose_branch x; AUTO_UPDATE="update"; inst_AdGuardHome "${AUTO_UPDATE:-update}"; } ; else end_op_message 1; fi
 			;;
+		"8" | "setipset")
+			PTXT "${INFO} IPSET integration imports compatible domain-to-IPSET mappings for routing and firewall add-ons." \
+				"${INFO} Disable it if this router does not use IPSET-based domain routing or filtering."
+			if read_yesno "Do you want to enable AdGuardHome IPSET integration?"; then check_ipset 1; else check_ipset 0; fi
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve that P2

**Comment by jumpsmm7** [comment]
@codex review

File: AdGuardHome.sh [hunk 5]
```diff
@@ -1725,12 +1723,17 @@ IPSet_Disable_Managed_For_Start_Locked() {
...
+}
+
 IPSet_Refresh() {
 	local RESTART_STATUS
+	IPSet_Enabled || return 0
 	IPSet_Supported || return 0
 	IPSET_REFRESH_CHANGED=""
 	IPSET_REFRESH_CONFIG="${1:-}"
-	IPSet_Lock IPSet_Refresh_Locked || return 1
+	IPSet_Lock IPSet_Setup_Locked || return 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh:1738]: @codex resolve that P1

File: installer [hunk 6]
```diff
@@ -1603,6 +1616,7 @@ setup_AdGuardHome_impl() {
 				if [ ! -f "${YAML_FILE}" ] || [ "$2" = "reconfig" ]; then
 					if read_yesno "Do you want to redirect all DNS resolutions on your network through to AdGuardHome?"; then check_dns_filter 1; else check_dns_filter 0; fi
 					if [ "$(nvram get dns_local_cache)" != "1" ] && read_yesno "Do you want to run AdGuardHome as a local caching DNS service which includes router traffic?"; then check_dns_local 1; else check_dns_local 0; fi
+					if read_yesno "Do you want to enable AdGuardHome IPSET integration?"; then check_ipset 1; else check_ipset 0; fi
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve that P2

**Comment by jumpsmm7** [comment]
@codex review for security regressions

File: AdGuardHome.sh [hunk 7]
```diff
@@ -770,14 +770,10 @@ start_adguardhome() {
...
 		if [ "${IPSET_START_STOPPED}" -eq 1 ] && IPSet_Start_Restore; then
 			IPSET_START_RESTARTED="1"
 		fi
-		if [ "${IPSET_START_RESTARTED}" -eq 1 ]; then
-			SERVICE_WAIT_TERMINAL_FAILURE="1"
-		fi
-		return 1
 	fi
 	if [ "${IPSET_START_RESTARTED}" -eq 0 ]; then
 		case "$(pidof "${PROCS}" 2>/dev/null | wc -w)" in
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh:781]: @codex address that P1

**Comment by jumpsmm7** [comment]
@codex review for security regressions

File: AdGuardHome.sh [hunk 8]
```diff
@@ -1783,6 +1788,20 @@ IPSet_Refresh_Locked() {
...
 	fi
 	rm -f "${RAW_TEMP_FILE}"
+	if ! awk '!/^[[:space:]]*(#|$)/ { found = 1; exit } END { exit !found }' "${TEMP_FILE}"; then
+		rm -f "${RAW_TEMP_FILE}" "${TEMP_FILE}"
+		IPSET_FILE_EXISTED=""
+		[ -e "${IPSET_FILE}" ] && IPSET_FILE_EXISTED="1"
+		if ! IPSet_Disable_Managed; then
+			return 1
+		fi
+		rm -f "${IPSET_FILE}"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply AdGuardHome.sh]: @codex resolve that P1

**Comment by jumpsmm7** [comment]
@codex review && @codex review for security regressions

File: installer [hunk 9]
```diff
@@ -1603,6 +1616,11 @@ setup_AdGuardHome_impl() {
 				if [ ! -f "${YAML_FILE}" ] || [ "$2" = "reconfig" ]; then
 					if read_yesno "Do you want to redirect all DNS resolutions on your network through to AdGuardHome?"; then check_dns_filter 1; else check_dns_filter 0; fi
 					if [ "$(nvram get dns_local_cache)" != "1" ] && read_yesno "Do you want to run AdGuardHome as a local caching DNS service which includes router traffic?"; then check_dns_local 1; else check_dns_local 0; fi
+					if read_yesno "Do you want to enable AdGuardHome IPSET integration?"; then IPSET_SELECTION=1; else IPSET_SELECTION=0; fi
+					if ! check_ipset "${IPSET_SELECTION}"; then
+						PTXT "${ERROR} Unable to save the AdGuardHome IPSET integration setting. Setup was aborted."
+						return 1
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve that P1

**Comment by jumpsmm7** [comment]
@codex review for security regressions

File: installer [hunk 10]
```diff
@@ -1603,6 +1620,14 @@ setup_AdGuardHome_impl() {
 				if [ ! -f "${YAML_FILE}" ] || [ "$2" = "reconfig" ]; then
 					if read_yesno "Do you want to redirect all DNS resolutions on your network through to AdGuardHome?"; then check_dns_filter 1; else check_dns_filter 0; fi
 					if [ "$(nvram get dns_local_cache)" != "1" ] && read_yesno "Do you want to run AdGuardHome as a local caching DNS service which includes router traffic?"; then check_dns_local 1; else check_dns_local 0; fi
+					if read_yesno "Do you want to enable AdGuardHome IPSET integration?"; then IPSET_SELECTION=1; else IPSET_SELECTION=0; fi
+					if ! check_ipset "${IPSET_SELECTION}"; then
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @codex resolve that P2

**Comment by jumpsmm7** [comment]
@codex review for security regressions

**Comment by jumpsmm7** [comment]
@codex review
